src/libinjector/linux/linux_utils.c

```diff
@@ -0,0 +1,250 @@
+    // we don't have injector_int3_userspace_cb function
+    // in scope here so we will use it from the previous trap
+    injector->bp->cb = info->trap->cb; //injector_int3_userspace_cb;
+    injector->bp->data = injector;
+    injector->bp->breakpoint.lookup_type = LOOKUP_DTB;
+    injector->bp->breakpoint.dtb = info->regs->cr3;
+    injector->bp->breakpoint.addr_type = ADDR_VA;
+    injector->bp->breakpoint.addr = syscall_addr + 2;
+    injector->bp->ttl = UNLIMITED_TTL;
+    injector->bp->ah_cb = NULL;
+
+    if ( drakvuf_add_trap(drakvuf, injector->bp) ) {
+        PRINT_DEBUG("Post syscall trap success\n");
+        return true;
+    }
+    else {
+        fprintf(stderr, "Couldn't trap next instruction after syscall\n");
+        return false;
+    }
+
+}
+
+bool save_rip_for_ret(drakvuf_t drakvuf, x86_registers_t* regs) {
+
+    // lock vmi
+    vmi_instance_t vmi = drakvuf_lock_and_get_vmi(drakvuf);
+
+    ACCESS_CONTEXT(ctx,
+                   .translate_mechanism = VMI_TM_PROCESS_DTB,
+                   .dtb = regs->cr3
+                  );
+    addr_t addr = regs->rsp;
+
+    // make space for storing rip
+    addr -= 0x8;
+    ctx.addr = addr;
+
+    if (VMI_FAILURE == vmi_write_64(vmi, &ctx, &regs->rip)) {
+        // release before returning
+        drakvuf_release_vmi(drakvuf);
+        return false;
+    }
+
+    regs->rsp = addr;
+
+    // release vmi
+    drakvuf_release_vmi(drakvuf);
+    return true;
+
+}
+
+bool load_file_to_injector_shellcode(injector_t injector, const char* file)
+{
+    FILE* fp = fopen(file, "rb");
+    if (!fp)
+        return false;
+
+    fseek (fp, 0, SEEK_END);
+    if ( (injector->shellcode.len = ftell (fp)) < 0 )
+    {
+        fclose(fp);
+        return false;
+    }
+    rewind (fp);
+
+    // we are adding +1 as we will append ret instruction for restoring the state of the VM
+    injector->shellcode.data = g_try_malloc0(injector->shellcode.len + 1);
+    if ( !injector->shellcode.data )
+    {
+        fclose(fp);
+        injector->shellcode.len = 0;
+        return false;
+    }
+
+    if ( (size_t)injector->shellcode.len != fread(injector->shellcode.data, 1, injector->shellcode.len, fp))
+    {
+        g_free(injector->shellcode.data);
+        injector->shellcode.data = NULL;
+        injector->shellcode.len = 0;
+        fclose(fp);
+        return false;
+    }
+    *(char *)(injector->shellcode.data + injector->shellcode.len ) = 0xc3; //ret
+    injector->shellcode.len += 1; // increase the length in variable
+
+    PRINT_DEBUG("Shellcode loaded to injector->shellcode\n");
+    print_hex(injector->shellcode.data, injector->shellcode.len, -1);
+
+    fclose(fp);
+
+    return true;
+}
+
+void free_injector(injector_t injector)
+{
+    if (!injector) return;
+
+    PRINT_DEBUG("Injector freed\n");
+
+    if (injector->cr3_trap) {
+        drakvuf_remove_trap(injector->drakvuf, injector->cr3_trap, NULL);
+        injector->cr3_trap = NULL;
+    }
```

Comment:
If I did not do this earlier, I used to get seg fault, but now when I checked in valgrind, I am getting some errors in it due to this